### 실습 1. 전처리 데이터 불러오기

해야 할 일

먼저 Chapter 01에서 만든 CSV 파일을 pandas로 불러옵니다.

처음부터 단어를 분리하지 말고 데이터가 정상적으로 연결되는지 먼저 확인합니다.

In [3]:
import pandas as pd

# CSV 파일 불러오기
df_books = pd.read_csv("book_bestseller_clean.csv", encoding="utf-8-sig")

# 1. 전체 행과 컬럼 개수
print("\n====shape:=======")
print(df_books.shape)

# 2. 컬럼 이름
print("\n====columns:=======")
print(df_books.columns)

# 3. 앞의 5행
print("\n====앞의 5행:=======")
display(df_books.head())

# 4. 상품명 컬럼 존재 여부
print("\n====상품명 컬럼 존재 여부:=======")
print("상품명" in df_books.columns)

shape: (199, 8)

====columns:=======
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

앞의 5행:


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설



상품명 컬럼 존재 여부:
True


### 실행 결과

- `book_bestseller_clean.csv` 파일을 `df_books` DataFrame으로 정상적으로 불러왔다.
- 데이터 크기는 **199행, 8컬럼**으로 확인되었다.
- 컬럼은 `순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`로 구성되어 있다.
- 앞의 5행 데이터가 정상적으로 출력되었다.
- `상품명` 컬럼 존재 여부를 확인한 결과 **True**로 확인되었다.

### 실습 2. 분석할 도서 제목 확인하기

단어 분석을 시작하기 전에 상품명 데이터가 어떤 형태인지 직접 확인합니다.

In [7]:
df_books["상품명"].head(20)
df_books["상품명"].isna().sum()
titles = (

    df_books["상품명"]
    .fillna("")
    .astype(str)
    .str.strip()

)

titles.head()


0        소년이 온다
1            모순
2    결국 국민이 합니다
3           혼모노
4            급류
Name: 상품명, dtype: str

In [8]:
import pandas as pd

# CSV 파일 불러오기
df_books = pd.read_csv("book_bestseller_clean.csv", encoding="utf-8-sig")

# 데이터 크기 확인
print("\n====shape:=======")
print(df_books.shape)

# 컬럼 이름 확인
print("\n====columns:=======")
print(df_books.columns)

# 앞의 5행 확인
print("\n====앞의 5행:=======")
display(df_books.head())

# 상품명 컬럼 존재 여부 확인
print("\n====상품명 컬럼 존재 여부:=======")
print("상품명" in df_books.columns)


====shape:=======
(199, 8)

====columns:=======
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

====앞의 5행:=======


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설



====상품명 컬럼 존재 여부:=======
True


### 상품명 데이터 탐색 결과 요약

- `book_bestseller_clean.csv` 파일을 정상적으로 불러왔다.
- 데이터 크기는 **199행, 8컬럼**으로 확인되었다.
- 분석에 필요한 `상품명` 컬럼이 포함되어 있다.
- 앞의 5행을 확인한 결과 상품명이 정상적으로 저장되어 있었다.
- `"상품명" in df_books.columns` 결과가 **True**로 확인되어, 이후 상품명 기반 단어 빈도 분석을 진행할 수 있는 상태이다.

### 실습 3. Dictionary로 단어 빈도 원리 이해하기 실습 전 이해하기
이번 실습은 여러 단어를 하나씩 확인하면서 **각 단어가 몇 번 나왔는지 세는 작업**이다.
같은 단어가 다시 나오면 기존 개수에 **1을 더하고**, 처음 나온 단어는 **1부터 시작**한다.
이 작업을 반복해서 최종적으로 **단어별 등장 횟수를 Dictionary에 저장**한다.

In [9]:
# 샘플 단어 목록
sample_words = [
    "데이터",
    "분석",
    "파이썬",
    "데이터",
    "AI",
    "분석",
    "데이터",
]

# 단어별 등장 횟수를 저장할 Dictionary
word_counts = {}

# 단어를 하나씩 확인하면서 개수 세기
for word in sample_words:
    word_counts[word] = word_counts.get(word, 0) + 1

# 결과 확인
print("\n====단어 빈도 결과:=======")
print(word_counts)


====단어 빈도 결과:=======
{'데이터': 3, '분석': 2, '파이썬': 1, 'AI': 1}


### 단어 빈도 계산 결과

- `sample_words`의 단어를 하나씩 반복하면서 등장 횟수를 계산했다.
- 결과는 `{'데이터': 3, '분석': 2, '파이썬': 1, 'AI': 1}`로 확인되었다.
- `데이터`는 3번, `분석`은 2번, `파이썬`과 `AI`는 각각 1번 등장했다.
- Dictionary를 이용하면 단어별 등장 횟수를 저장하고 확인할 수 있다.

### 실습 4. Counter로 단어 빈도 계산하기

이번 실습에서는 `Counter`를 사용해 단어별 등장 횟수를 더 간단하게 계산한다.

- `Counter(sample_words)` → 각 단어의 빈도를 자동으로 계산
- `.most_common()` → 많이 나온 단어부터 확인
- `.most_common(3)` → 상위 3개 단어만 확인

Dictionary는 빈도 계산 원리를 직접 이해하기 좋고,  
Counter는 같은 작업을 더 짧고 편리하게 처리할 때 사용한다.

In [10]:
from collections import Counter

# 단어 빈도 계산
sample_counter = Counter(sample_words)

# 전체 단어 빈도 확인
print("\n====전체 단어 빈도:=======")
print(sample_counter)

# 많이 나온 순서대로 확인
print("\n====빈도순 정렬:=======")
print(sample_counter.most_common())

# 상위 3개 단어 확인
print("\n====상위 3개 단어:=======")
print(sample_counter.most_common(3))


====전체 단어 빈도:=======
Counter({'데이터': 3, '분석': 2, '파이썬': 1, 'AI': 1})

====빈도순 정렬:=======
[('데이터', 3), ('분석', 2), ('파이썬', 1), ('AI', 1)]

====상위 3개 단어:=======
[('데이터', 3), ('분석', 2), ('파이썬', 1)]


### Counter 단어 빈도 계산 결과

- `Counter`를 사용해 각 단어의 등장 횟수를 자동으로 계산했다.
- 전체 결과는 `데이터 3회`, `분석 2회`, `파이썬 1회`, `AI 1회`로 확인되었다.
- `most_common()`을 사용하면 등장 횟수가 많은 순서대로 정렬할 수 있다.
- `most_common(3)`으로 가장 많이 등장한 상위 3개 단어만 확인할 수 있다.

### 실습 5. pandas `value_counts()`로 빈도 확인하기

이번 실습에서는 pandas의 `value_counts()`를 이용해 각 단어가 몇 번 등장했는지 확인한다.

- `pd.Series(sample_words)` → 단어 목록을 Series로 변환
- `.value_counts()` → 같은 값이 몇 번 나왔는지 자동으로 계산
- 결과는 `Counter`로 계산한 빈도와 비슷하다.

### 세 방법 비교

| 방법 | 특징 | 의미 |
|---|---|---|
| Dictionary | 직접 빈도 누적 | 원리 이해 |
| Counter | Python 표준 기능 | 간단한 빈도 계산 |
| value_counts() | pandas 기능 | Series 기반 빈도 분석 |

분석 상황에 따라 가장 편한 방법을 선택해서 사용할 수 있다.

In [11]:
import pandas as pd

# 단어 목록을 Series로 변환
sample_series = pd.Series(sample_words)

# 단어별 빈도 확인
print("\n====value_counts 결과:=======")
print(sample_series.value_counts())


====value_counts 결과:=======
데이터    3
분석     2
파이썬    1
AI     1
Name: count, dtype: int64


### value_counts() 단어 빈도 확인 결과

- `sample_words`를 pandas `Series`로 변환했다.
- `value_counts()`를 사용해 단어별 등장 횟수를 계산했다.
- 결과는 `데이터 3회`, `분석 2회`, `파이썬 1회`, `AI 1회`로 확인되었다.
- `Counter`와 비슷한 결과를 pandas에서도 간단하게 확인할 수 있었다.

### Counter와 value_counts() 비교

- 일반 Python 리스트 데이터 → `Counter`
- pandas의 DataFrame / Series 데이터 → `value_counts()`
- 대량 데이터를 pandas로 분석할 때는 `value_counts()`가 더 편리한 경우가 많다.

이번 교보문고 프로젝트에서는 pandas를 사용하고 있으므로 `value_counts()`를 사용하면 된다.

### 실습 6. 도서 제목을 간단하게 단어로 나누기

이번 실습에서는 실제 `상품명`을 정규표현식으로 간단하게 나누어 단어 목록을 만든다.

- 한글, 영어, 숫자만 추출해서 토큰으로 나눈다.
- 영어는 소문자로 바꿔서 같은 단어를 통일한다.
- 이렇게 만든 단어를 이용해 이후 빈도 분석을 할 수 있다.
- 다만 `시대의`, `python으로`처럼 의미 단위가 정확히 나뉘지 않을 수 있어, 이후 형태소 분석이 필요하다.

핵심 흐름은  
`상품명 → 단어로 나누기 → 단어 빈도 계산` 이다.

In [ ]:
# 정규표현식 기능을 사용하기 위해 re 모듈 불러오기
import re

# 문자열을 간단하게 단어로 나누는 함수 만들기
def simple_tokenize(text):

    # 1. 입력된 값을 문자열로 바꾸기
    text = str(text)

    # 2. 영어 대문자를 소문자로 바꾸기
    text = text.lower()

    # 3. 한글, 영어, 숫자만 찾아서 단어 목록으로 만들기
    words = re.findall(r"[가-힣A-Za-z0-9]+", text)

    # 4. 추출한 단어 목록 반환하기
    return words


# 테스트할 문장 준비
sample_title = "AI 시대의 데이터 분석, Python으로 시작하기!"

# 함수에 문장을 넣어서 단어로 나누기
result = simple_tokenize(sample_title)

# 결과 확인
print("\n====토큰화 결과:=======")
print(result)


====토큰화 결과:=======
['ai', '시대의', '데이터', '분석', 'python으로', '시작하기']


### 실습 7. 모든 제목에서 단어 추출하기

이번 실습에서는 모든 도서 제목을 하나씩 확인하면서 단어를 추출하고,  
추출한 단어들을 `simple_words` 리스트 하나에 모두 모은다.

- `simple_tokenize()` → 제목을 단어로 나누기
- `extend()` → 추출된 단어들을 기존 리스트에 하나씩 이어 붙이기
- `len(simple_words)` → 전체 단어 수 확인
- `simple_words[:50]` → 앞의 50개 단어 확인

`append()`는 리스트 안에 리스트가 들어가지만,  
`extend()`는 단어들을 하나의 리스트에 이어 붙이기 때문에 빈도 분석에 더 적합하다.

### append()와 extend() 차이
둘 다 리스트에 값을 추가할 때 사용하지만, 추가되는 모양이 다르다.

| 방법 | 동작 | 결과 |
|---|---|---|
| `append()` | 값을 하나의 덩어리로 추가 | 리스트 안에 리스트가 들어갈 수 있음 |
| `extend()` | 여러 값을 하나씩 펼쳐서 추가 | 기존 리스트에 값들이 이어 붙음 |

### 예시
```python
example = []

example.append(["데이터", "분석"])
print(example)

In [14]:
# 상품명 컬럼 가져오기
titles = df_books["상품명"]

# 추출한 모든 단어를 저장할 빈 리스트 만들기
simple_words = []

# 상품명을 하나씩 반복해서 확인
for title in titles:
    # 제목 1개를 단어로 나누기
    words = simple_tokenize(title)

    # 나눈 단어들을 simple_words에 하나씩 추가
    simple_words.extend(words)

# 전체 단어 수 확인
print("\n====전체 단어 수:=======")
print(len(simple_words))

# 앞의 50개 단어 확인
print("\n====앞의 50개 단어:=======")
print(simple_words[:50])


====전체 단어 수:=======
760

====앞의 50개 단어:=======
['소년이', '온다', '모순', '결국', '국민이', '합니다', '혼모노', '급류', '초역', '부처의', '말', '청춘의', '독서', '특별증보판', '어른의', '행복은', '조용하다', '채식주의자', '단', '한', '번의', '삶', '강물에디션', '활판인쇄', '한정판', '작별하지', '않는다', '스토너', '해커스', '토익', '기출', 'voca', '보카', '다크', '심리학', '20만', '부', '한정판', '스페셜', '에디션', '트렌드', '코리아', '2026', '위버멘쉬', '가공범', '듀얼', '브레인', '행동하지', '않으면', '인생은']


### 모든 상품명에서 단어 추출 결과

- `상품명` 컬럼의 제목을 하나씩 확인하고 `simple_tokenize()`로 단어를 나눴다.
- `extend()`를 사용해 추출한 단어들을 하나의 리스트에 모았다.
- 전체 단어 수는 **760개**로 확인되었다.
- 앞의 단어를 확인해 보니 `소년이`, `국민이`, `부처의`, `청춘의`처럼 조사나 어미가 붙은 단어도 함께 추출되었다.
- 즉, 단순 토큰화만으로는 사람이 생각하는 정확한 의미 단위로 나누기 어렵다는 점을 확인했다.
- 이런 한계 때문에 이후에는 형태소 분석을 이용해 단어를 더 정확하게 나눌 필요가 있다.

### 실습 8. 실제 데이터에서 단어 빈도 확인하기

이번 실습에서는 모든 상품명에서 추출한 단어를 `Counter`로 집계해서  
가장 많이 등장한 단어를 확인한다.

- `Counter(simple_words)` → 단어별 등장 횟수 계산
- `most_common(30)` → 자주 나온 상위 30개 단어 확인
- pandas DataFrame으로 바꿔서 `단어`, `빈도` 형태의 표로 확인

바로 Word Cloud를 만들기보다 상위 단어를 먼저 확인하고,  
불필요한 단어·숫자·조사·중복 표현이 있는지 보고 다음 전처리 방향을 결정한다.

In [15]:
from collections import Counter
import pandas as pd

# simple_words에 있는 단어들의 등장 횟수 계산
simple_counter = Counter(simple_words)

# 가장 많이 나온 상위 30개 단어 확인
print("\n====상위 30개 단어:=======")
print(simple_counter.most_common(30))

# 상위 30개 단어를 표 형태로 만들기
simple_top30 = pd.DataFrame(
    simple_counter.most_common(30),
    columns=["단어", "빈도"]
)

# 표로 확인
print("\n====상위 30개 단어 표:=======")
display(simple_top30)


====상위 30개 단어:=======
[('에디션', 12), ('부', 9), ('기념', 8), ('리커버', 8), ('한', 7), ('해커스', 7), ('위한', 7), ('말', 6), ('토익', 6), ('스페셜', 5), ('나는', 5), ('내가', 5), ('대하여', 4), ('싶은', 4), ('lc', 4), ('rc', 4), ('1', 4), ('단', 3), ('한정판', 3), ('않는다', 3), ('voca', 3), ('트렌드', 3), ('2026', 3), ('30만', 3), ('돈의', 3), ('때', 3), ('2025', 3), ('일본어', 3), ('없는', 3), ('필사', 3)]

====상위 30개 단어 표:=======


,단어,빈도
0,에디션,12
1,부,9
2,기념,8
3,리커버,8
4,한,7
5,해커스,7
6,위한,7
7,말,6
8,토익,6
9,스페셜,5


### 상위 단어 빈도 확인 결과

- `Counter`로 전체 단어의 빈도를 계산한 뒤 상위 30개를 확인했다.
- 가장 많이 나온 단어는 `에디션` 12회였고, `부` 9회, `기념`과 `리커버`가 각각 8회로 나타났다.
- `한`, `위한`, `나는`, `내가`, `돈의`처럼 조사나 어미가 붙은 표현도 많이 포함되어 있었다.
- `1`, `2025`, `2026`, `30만`처럼 숫자 표현도 함께 추출되었다.
- 따라서 현재 결과에는 분석에 바로 쓰기 어려운 단어가 섞여 있어, 다음 단계에서 불필요한 단어 제거와 형태소 분석이 필요하다.
### 정규표현식 토큰화 상위 30개 단어 분석 결과
정규표현식을 이용해 도서 제목을 단어 단위로 나눈 뒤, `Counter`를 사용해 상위 30개 단어의 빈도를 확인했다.
가장 많이 등장한 단어는 `에디션` 12회였으며, `부` 9회, `기념` 8회, `리커버` 8회 순으로 나타났다. 또한 `해커스`, `토익`, `트렌드`, `필사`, `일본어`처럼 실제 도서 주제나 분야를 나타내는 단어도 확인할 수 있었다.
반면 `한`, `위한`, `나는`, `내가`, `돈의`, `없는`처럼 조사나 어미가 붙은 표현과 `1`, `2025`, `2026`, `30만` 같은 숫자 표현도 상위 빈도에 포함되었다.
즉, 정규표현식만으로도 전체적인 단어 빈도는 확인할 수 있지만, 의미 있는 핵심 단어만 정확하게 추출하기에는 한계가 있다. 이후 분석에서는 불필요한 표현을 제거하고 형태소 분석을 적용해 단어를 더 정제할 필요가 있다.

### 실습 9. pandas로 실제 단어 빈도 확인하기

이번 실습에서는 앞에서 만든 `simple_words`를 pandas `Series`로 변환한 뒤,  
`value_counts()`를 사용해 단어별 빈도를 확인한다.

- `pd.Series(simple_words)` → 단어 리스트를 Series로 변환
- `value_counts().head(30)` → 자주 나온 상위 30개 단어 확인
- `Counter` 결과와 비교해서 상위 단어와 빈도가 같은지 검증

두 결과가 같다면 단어 추출과 빈도 계산이 정상적으로 진행된 것으로 볼 수 있다.

In [16]:
# 단어 리스트를 pandas Series로 변환
simple_word_series = pd.Series(simple_words)

# pandas로 상위 30개 단어 확인
print("\n====pandas 상위 30개 단어:=======")
print(simple_word_series.value_counts().head(30))

# Counter와 pandas 결과 비교
print("\n====Counter 상위 10개:=======")
print(simple_counter.most_common(10))

print("\n====pandas 상위 10개:=======")
print(simple_word_series.value_counts().head(10))


====pandas 상위 30개 단어:=======
에디션     12
부        9
기념       8
리커버      8
한        7
해커스      7
위한       7
말        6
토익       6
스페셜      5
나는       5
내가       5
대하여      4
싶은       4
lc       4
rc       4
1        4
단        3
한정판      3
않는다      3
voca     3
트렌드      3
2026     3
30만      3
돈의       3
때        3
2025     3
일본어      3
없는       3
필사       3
Name: count, dtype: int64

====Counter 상위 10개:=======
[('에디션', 12), ('부', 9), ('기념', 8), ('리커버', 8), ('한', 7), ('해커스', 7), ('위한', 7), ('말', 6), ('토익', 6), ('스페셜', 5)]

====pandas 상위 10개:=======
에디션    12
부       9
기념      8
리커버     8
한       7
해커스     7
위한      7
말       6
토익      6
스페셜     5
Name: count, dtype: int64


### pandas 단어 빈도 확인 결과

- `simple_words`를 pandas `Series`로 변환한 뒤 `value_counts()`로 단어 빈도를 확인했다.
- 상위 단어는 `에디션` 12회, `부` 9회, `기념` 8회, `리커버` 8회 등으로 나타났다.
- `Counter`에서 확인한 상위 단어와 빈도도 같은 결과로 확인되었다.
- 따라서 `Counter`와 `value_counts()`가 같은 단어 빈도를 계산하고 있음을 확인했다.
- 다만 `한`, `위한`, `나는`, `내가`처럼 의미 분석에 바로 쓰기 어려운 표현도 포함되어 있어 추가 전처리가 필요하다.

### 실습 10. 왜 형태소 분석이 필요한가?
한국어는 조사와 어미가 붙기 때문에 단순히 공백이나 정규표현식만으로 단어를 정확히 나누기 어렵다.
예를 들어 `데이터를`, `데이터가`, `데이터의`, `데이터로`는 모두 핵심 단어가 `데이터`이지만 서로 다른 문자열로 처리될 수 있다.
또한 `은`, `는`, `이`, `가`, `을`, `를`, `의` 같은 조사도 함께 포함될 수 있다.
따라서 이번 실습에서는 형태소 분석을 이용해 도서 제목에서 **명사 중심의 핵심 단어를 추출**한다.

In [17]:
# 형태소 분석기 불러오기
from konlpy.tag import Okt

# 형태소 분석기 준비
okt = Okt()

# 테스트 문장
sample_text = "데이터를 데이터가 데이터의 데이터로 분석합니다."

# 명사만 추출
nouns = okt.nouns(sample_text)

# 결과 확인
print("\n====명사 추출 결과:=======")
print(nouns)

ModuleNotFoundError: No module named 'konlpy'